### PR TITLE
🐛 Create a connection in the acceptance test setup and close is in the tear down

### PR DESCRIPTION
**Pull request by benmoriceau**
## What
Update the Redshift acceptance test in order to limit the number of connections opened by the redshift acceptance tests.
We are now directly creating a connection without using a datasource. It is needed because it has been observed that:
- `datasource.close()` doesn't close the opened connection
- According to the javadoc, `datasource.getConnection()` may return a new connection.
This means that we can not ensure that a connection is being closed by using the datasource. Which explains why we switch to using a Connection directly.

## Recommended reading order
1. `ConnectionFactory.java`
2. The test files.

### CODE REVIEW

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java
Comment (tryangul) [inline airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java:31]: final

Comment (benmoriceau) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java]: Thanks! I wonder why my plugin didn't catch it.

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3DestinationAcceptanceTest.java
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3DestinationAcceptanceTest.java:130]: nit: I think this can be one line, not sure why the this wasn't previously the case since L137, L147 are the same.

**Review by evantahler — COMMENTED** [review]
The tests actually pass, so 👍 from me!

I'd want another review (paging @jbfbell) to answer the question of if this is a special thing we are OK with for just redshift, of we should make this kind of change for all JDBC destination tests

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/SshRedshiftDestinationBaseAcceptanceTest.java
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/SshRedshiftDestinationBaseAcceptanceTest.java:224]: +1

Comment (evantahler) [reply airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/SshRedshiftDestinationBaseAcceptanceTest.java:224]: Can these 3 commands be wrapped in another method so we don't forget to do them all?

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java
Comment (davinchia) [inline airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java:23]: nit: can we also add to the javadoc that we need this because we run tests in parallel and the usual data source does not have sufficient connection guarantees to prevent creating too many connections

**Review by davinchia — COMMENTED** [review]
Great stuff @benmoriceau ! 

Makes sense to me. I will defer to the Destinations team for approval merge.

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3DestinationAcceptanceTest.java
Comment (benmoriceau) [reply airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3DestinationAcceptanceTest.java:130]: Done

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/SshRedshiftDestinationBaseAcceptanceTest.java
Comment (benmoriceau) [reply airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/SshRedshiftDestinationBaseAcceptanceTest.java:224]: Done

**Comment by benmoriceau** [comment]
> The tests actually pass, so 👍 from me!
> 
> I'd want another review (paging @jbfbell) to answer the question of if this is a special thing we are OK with for just redshift, of we should make this kind of change for all JDBC destination tests

I looked a Snowflake as well. One big difference is that Snowflake is creating one connection per test where we were creating one connection per query for redshift.

I haven't check the implementation of the redshiftJDBC connection but it could come from it as well, the javadoc says that the behavior of close on an un-commited connection is unknown.

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java
Comment (benmoriceau) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java:23]: Added a comment on the class itself

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java
Comment (davinchia) [inline airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java:16]: What does 'shouldn't be needed in our application code' mean?

Should we replace DataSource through database code base with this?

**Review by davinchia — APPROVED** [review]
Great stuff!

Just one minor clarification.

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java
Comment (benmoriceau) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java:16]: This factory, I mean that we can use a datasource directly. Since it is created using Micronaut, it will be a singleton and thus reduce the leak risk. Changing that in our applications will make use to do unnecessary changes knowing that the datasource is not even created by use and managed through the application.yaml. It will need us to re-implement some options that we are using.

Here is an example of a datasource declaration from the worker:
```
datasources:
  config:
    connection-test-query: SELECT 1
    connection-timeout: 30000
    idle-timeout: 600000
    initialization-fail-timeout: -1 # Disable fail fast checking to avoid issues due to other pods not being started in time
    maximum-pool-size: 20
    minimum-idle: 0
    url: ${DATABASE_URL}
    driverClassName: org.postgresql.Driver
    username: ${DATABASE_USER}
    password: ${DATABASE_PASSWORD}
    pool-name: config-pool
```

**Review by jbfbell — APPROVED** [review]

**Comment by octavia-approvington** [comment]
Lets do it! <br/> giddddy-up

**Review by octavia-approvington — APPROVED** [review]

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java
Comment (davinchia) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/ConnectionFactory.java:16]: In that case, can we clarify with something like, this does not apply to DI frameworks e.g. Micronaut?

'Application code' is confusing because there is a wide breadth of application code at Airbyte. A reader is going to get confused if they are going between Platform (Micronaut-ed) and Connectors (not-micronauted)

**Pull request merged**
